content/struct-classes.md

```diff
@@ -0,0 +1,627 @@
+
+### Compatibility
+
+The changes described in this proposal are backward binary compatible because they do
+not affect the bytecode produced by the existing language features.
+
+The changes described in this proposal are backward TASTy compatible because they
+don’t require any changes at the TASTy level.
+
+Source compatibility is preserved, and the semantics of existing valid programs is not
+changed because the proposal does not change the existing language features, it only
+adds a new keyword, `struct`.
+
+### Open questions
+
+#### Choice of keyword `struct`
+
+The keyword `struct` stands for “structural”, which relates directly to the semantics
+of `struct class` definitions: class definitions that automatically implement
+structural equality (as opposed to the default object reference equality).
+
+However, `struct` has historically also been used in C and C++ to define “data structures”,
+with no relation with structural equality. The keyword `struct` is also used in Swift and
+Rust with a similar meaning (“data structure”).
+
+Other languages implementing a similar feature use the following keywords:
+- `record` (Java), a restricted variant of `struct class` (Java records can’t
+  extend parent classes, can’t be extended, and can’t have mutable fields --- even
+  private ones),
+- `data class` (Kotlin), which are equivalent to `case classes` in Scala and have
+  the same issues regarding binary compatibility.
+
+We think that using the keyword `record` or `data` instead of `struct` would bring
+confusion because the feature these keywords relate to in other programming languages
+is closer to Scala `case classes` than `struct classes`. On the other hand, we agree
+that in other programming languages the keyword `struct` may have an established
+meaning that is different from the one we give here, so we are open to using a
+different keyword, such as `structural`.
+
+## Alternatives
+
+### Automatic generation of transformation methods
+
+Besides the two solutions shown in the Motivation section (based on regular classes
+or case classes), we also considered a more powerful variant of `struct class` that
+would also automatically generate the transformation methods (`withName`, `withAge`, 
+etc. in the example). However, this solution is more complex to specify, and it also
+has some annoying drawbacks. In particular, it does not work well with all the names,
+especially with symbolic names (e.g., `+`, `%`, etc.), or names that contain
+acronyms (e.g. consider a field `httpHeaders`, should the transformation method
+be named `withHttpHeaders` or `withHTTPHeaders`?). Furthermore, it would be extremely
+complicated to specify rules to handle "smarter" transformation methods, such as
+`withoutEmail` in our example, to specifically handle fields whose type is `Option`,
+or a collection.
+
+### Leverage `Product`
+
+We could introduce a new trait, say `Structural`, to the standard library,
```

Comment:
Would it be a good idea from then on to have all case classes extending specifically `Structural` instead of the more general `Product`?
Aren't _case classes_ specific cases of `Structural`?

Reply:
That makes sense, yes. This is something we could discuss further in case the marker trait approach is selected.